pyteal/ast/itxn.py

```diff
@@ -200,6 +206,111 @@ def SetFields(cls, fields: Dict[TxnField, Union[Expr, List[Expr]]]) -> Expr:
         fieldsToSet = [cls.SetField(field, value) for field, value in fields.items()]
         return Seq(fieldsToSet)
 
+    @classmethod
+    def MethodCall(
+        cls,
+        *,
+        app_id: Expr,
+        method_signature: str,
+        args: List[Union[abi.BaseType, Expr, Dict[TxnField, Union[Expr, List[Expr]]]]],
+        fields: Dict[TxnField, Union[Expr, List[Expr]]],
+    ) -> Expr:
+
+        # Default, always need these
+        fields_to_set = [
+            cls.SetField(TxnField.type_enum, TxnType.ApplicationCall),
+            cls.SetField(TxnField.application_id, app_id),
+        ]
+
+        # We only care about the args
+        arg_type_specs, _ = type_specs_from_signature(method_signature)
+
+        # Start app args with the method selector
+        app_args: List[Expr] = [MethodSignature(method_signature)]
+
+        # Transactions are not included in the App Call
+        txns_to_pass: List[Expr] = []
+
+        # Reference Types are treated specially
+        accts: List[Expr] = []
+        apps: List[Expr] = []
+        assets: List[Expr] = []
+
+        # Reference types are encoded as a uint8 index into
+        # their respective arrays
+        def get_uint8(i):
+            return Suffix(Itob(Int(i)), Int(7))
+
+        for idx, arg_ts in enumerate(arg_type_specs):
+            arg = args[idx]
+
+            if arg_ts in abi.TransactionTypeSpecs:
+                if not isinstance(arg, Dict):
+                    raise TealTypeError(arg, Dict[TxnField, Union[Expr, List[Expr]]])
+
+                txns_to_pass.append(InnerTxnBuilder.SetFields(arg))
+
+            elif arg_ts in abi.ReferenceTypeSpecs:
+                match arg_ts:
+                    # For both acct and application, add index to
+                    # app args _after_ appending since 0 is implicitly set
+                    case abi.AccountTypeSpec():
+                        if isinstance(arg, Expr):
+                            accts.append(arg)
+                        elif isinstance(arg, abi.Account):
+                            accts.append(arg.address())
+                        else:
+                            raise TealTypeError(arg, Union[abi.Account, Expr])
+
+                        app_args.append(get_uint8(len(accts)))
+
+                    case abi.ApplicationTypeSpec():
+                        if isinstance(arg, Expr):
+                            apps.append(arg)
+                        elif isinstance(arg, abi.Application):
+                            apps.append(arg.application_id())
+                        else:
+                            raise TealTypeError(arg, Union[abi.Application, Expr])
+
+                        app_args.append(get_uint8(len(apps)))
+
+                    # For assets, add to app_args prior to appending to assets array
+                    case abi.AssetTypeSpec():
+                        app_args.append(get_uint8(len(assets)))
```

Comment:
Nice ARC-4 nuance to treat `assets` index differently from `accounts` and `applications` !!!